### PR TITLE
Customize metadata with scripting api

### PR DESCRIPTION
**Pull request by GRL78**
Merge #1213 + gameMap manage the phaser layer

### CODE REVIEW

**Comment by moufmouf** [comment]
There are a lot of data related to the current user. We should probably put those in a dedicated object.

Something like:

`WA.getCurrentUser(): Promise<User>`

and

```
interface User {
    id: string,
    nickName: string,
    tags: string[]
}
```

File: front/src/Api/Events/ChangeTileEvent.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+export const isChangeTileEvent = tg.isArray(
+    new tg.IsInterface().withProperties({
+        x: tg.isNumber,
+        y: tg.isNumber,
+        tile: tg.isUnion(tg.isNumber, tg.isString),
+        layer: tg.isUnion(tg.isNumber, tg.isString)
+    }).get()
+);
+/**
+ * A message sent from the game to the iFrame when a user enters or leaves a zone marked with the "zone" property.
```
Comment (moufmouf) [inline front/src/Api/Events/ChangeTileEvent.ts]: Fix this comment

File: front/src/Api/Events/GameStateEvent.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+export const isGameStateEvent =
+    new tg.IsInterface().withProperties({
+        roomId: tg.isString,
+        mapUrl: tg.isString,
+        nickname: tg.isUnion(tg.isString, tg.isNull),
+        uuid: tg.isUnion(tg.isString, tg.isUndefined),
+        startLayerName: tg.isUnion(tg.isString, tg.isNull)
+    }).get();
+/**
+ * A message sent from the game to the iFrame when the gameState is got by the script
```
Comment (moufmouf) [inline front/src/Api/Events/GameStateEvent.ts:26]:
```suggestion
 * A message sent from the game to the iFrame when the gameState is received by the script
```

File: front/src/Api/Events/HasPlayerMovedEvent.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+export const isHasPlayerMovedEvent =
+    new tg.IsInterface().withProperties({
+        direction: tg.isString,
+        moving: tg.isBoolean,
+        x: tg.isNumber,
+        y: tg.isNumber
+    }).get();
+
+/**
+ * A message sent from the game to the iFrame when the player move after the iFrame send a message to the game that it want to listen to the position of the player
```
Comment (moufmouf) [inline front/src/Api/Events/HasPlayerMovedEvent.ts:14]:
```suggestion
 * A message sent from the game to the iFrame to notify a movement from the current player.
```

File: front/src/Api/Events/HasPlayerMovedEvent.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import * as tg from "generic-type-guard";
+
+
+
+export const isHasPlayerMovedEvent =
+    new tg.IsInterface().withProperties({
+        direction: tg.isString,
```
Comment (moufmouf) [inline front/src/Api/Events/HasPlayerMovedEvent.ts:7]: You can probably express direction as a list of 4 possible values only.

File: front/src/Api/Events/TagEvent.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import * as tg from "generic-type-guard";
+
+export const isTagEvent =
+    new tg.IsInterface().withProperties({
+        list: tg.isArray(tg.isString),
+    }).get();
+/**
+ * A message sent from the iFrame to the game to show/hide a layer.
```
Comment (moufmouf) [inline front/src/Api/Events/TagEvent.ts:8]: Fix this.
Also, I think we don't need a specific event for this. We could use the "GameStateEvent" to fetch tags.

File: front/src/Api/Events/TilesetEvent.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    new tg.IsInterface().withProperties({
+        name : tg.isString,
+        imgUrl : tg.isString,
+        tilewidth : tg.isNumber,
+        tileheight : tg.isNumber,
+        margin : tg.isNumber,
+        spacing : tg.isNumber,
+    }).get();
+/**
+ * A message sent from the iFrame to the game to show/hide a layer.
```
Comment (moufmouf) [inline front/src/Api/Events/TilesetEvent.ts:13]: Remove this (for now)

File: front/src/Connexion/RoomConnection.ts [hunk 7]
```diff
@@ -170,10 +169,10 @@ export class RoomConnection implements RoomConnection {
             } else if (message.hasWorldfullmessage()) {
                 worldFullMessageStream.onMessage();
                 this.closed = true;
-            } else if (message.hasWorldconnexionmessage()) {
-                worldFullMessageStream.onMessage(message.getWorldconnexionmessage()?.getMessage());
-                this.closed = true;
-            }else if (message.hasWebrtcsignaltoclientmessage()) {
+                // // } else if (message.hasWorldconnexionmessage()) {
+                //     worldFullMessageStream.onMessage(message.getWorldconnexionmessage()?.getMessage());
+                //      this.closed = true;
```
Comment (moufmouf) [inline front/src/Connexion/RoomConnection.ts:174]: Why is this code commented? It seems unrelated to the PR.

File: front/src/Connexion/RoomConnection.ts [hunk 8]
```diff
@@ -599,4 +598,11 @@ export class RoomConnection implements RoomConnection {
     public isAdmin(): boolean {
         return this.hasTag('admin');
     }
+
+    public getAllTag() : string[] {
```
Comment (moufmouf) [inline front/src/Connexion/RoomConnection.ts:602]: Remove this test method.

File: front/src/Phaser/Game/GameMap.ts [hunk 9]
```diff
@@ -100,4 +115,46 @@ export class GameMap {
...
+               found = true;
+            }
+            else {
+                i++;
+            }
+        }
+        if (found) {
+            return this.flatLayers[i];
+        }
+        return undefined;
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameMap.ts:133]: Can't we simply express this as:

```suggestion
    return this.flatLayers.find((layer) => layer.name = layerName);
```

?

File: front/src/Phaser/Game/GameMap.ts [hunk 10]
```diff
@@ -100,4 +115,46 @@ export class GameMap {
...
+                found = true;
+            }
+            else {
+                i++;
+            }
+        }
+        if (found) {
+            return this.phaserLayers[i];
+        }
+        return undefined;
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameMap.ts:150]: Idem, can't we express this as:

```suggestion
    return this.phaserLayers.find((layer) => layer.layer.name = layerName);
```

?

File: front/src/Phaser/Game/GameMap.ts [hunk 11]
```diff
@@ -100,4 +115,46 @@ export class GameMap {
...
+    }
+
+    public addTerrain(terrain : Phaser.Tilemaps.Tileset): void {
+        console.log('Add');
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameMap.ts:154]:
```suggestion
```

File: front/src/Phaser/Game/GameScene.ts [hunk 12]
```diff
@@ -837,53 +830,121 @@ ${escapedMessage}
...
             this.add.existing(scriptedBubbleSprite);
         }));
 
-       this.iframeSubscriptionList.push(iframeListener.removeBubbleStream.subscribe(()=>{
+        this.iframeSubscriptionList.push(iframeListener.removeBubbleStream.subscribe(() => {
             scriptedBubbleSprite.destroy();
         }));
 
+        this.iframeSubscriptionList.push(iframeListener.showLayerStream.subscribe((layerEvent)=>{
+            console.log('show');
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameScene.ts:889]:
```suggestion
```

File: front/src/Phaser/Game/GameScene.ts [hunk 13]
```diff
@@ -837,53 +830,121 @@ ${escapedMessage}
...
+            this.setLayerVisibility(layerEvent.name, true);
+        }));
+
+        this.iframeSubscriptionList.push(iframeListener.hideLayerStream.subscribe((layerEvent)=>{
+            console.log('hide');
```
Comment (moufmouf) [inline front/src/Phaser/Game/GameScene.ts:894]:
```suggestion
```

File: front/src/iframe_api.ts [hunk 14]
```diff
@@ -73,8 +101,126 @@ class Popup {
...
+function postToParent(content: IframeEvent<keyof IframeEventMap>) {
+    window.parent.postMessage(content, "*")
+}
+let playerUuid: string | undefined;
 
 window.WA = {
+
+    onPlayerMove(callback: HasPlayerMovedEventCallback): void {
+        playerUuid = uuidv4();
+        callbackPlayerMoved[playerUuid] = callback;
```
Comment (moufmouf) [inline front/src/iframe_api.ts:161]: Why do we need to playerUuid? Can't we simply add the callback to a simple list?

**Review by moufmouf — CHANGES_REQUESTED** [review]

File: front/src/iframe_api.ts (hunk 14 above)
Comment (jonnytest1) [reply front/src/iframe_api.ts:161]: i think that was sort of going in the direction of this 
https://github.com/thecodingmachine/workadventure/issues/1059
should be unneccessay in this commit though

File: CHANGELOG.md [hunk 15]
```diff
@@ -48,6 +48,13 @@
     - Added a new `DISPLAY_TERMS_OF_USE` environment variable to trigger the display of terms of use  
 - New scripting API features:
   - Use `WA.loadSound(): Sound` to load / play / stop a sound
+  - Use `WA.showLayer(): void` to show a layer
+  - Use `WA.hideLayer(): void` to hide a layer
+  - Use `WA.setProperty() : void` to add or change existing property of a layer
```
Comment (jonnytest1) [inline CHANGELOG.md:53]: the new method should be on a subobject (maybe a new one ) since the latest api refactor

File: front/src/Api/iframe/nav.ts [hunk 16]
```diff
@@ -31,7 +32,7 @@ class WorkadventureNavigationCommands extends IframeApiContribution<Workadventur
             "type": 'loadPage',
             "data": {
                 url
-            }
+            } as LoadPageEvent
```
Comment (jonnytest1) [inline front/src/Api/iframe/nav.ts:35]: its not strictly neccessary to type those objects as sendToWorkadventure in combination with the the event type already implicitly expects the right type

File: CHANGELOG.md (hunk 15 above)
Comment (GRL78) [reply CHANGELOG.md:53]: Currently working on it.

File: docs/maps/api-ui.md [hunk 17]
```diff
@@ -65,3 +65,25 @@ WA.room.onLeaveZone('myZone', () => {
...
+
+WA.ui.registerMenuCommand("test", () => {
+    WA.chat.sendChatMessage("test clicked", "menu cmd")
+})
+
+```
+
+<div class="col">
+    <img src="./assets/menu-command.png" class="figure-img img-fluid rounded" alt="" />
+</div>
```
Comment (moufmouf) [inline docs/maps/api-ui.md:89]: This will probably not work in the website. We should hard code a "raw" content URL (alas)

Comment (jonnytest1) [reply docs/maps/api-ui.md:89]: actually :3 https://github.com/thecodingmachine/workadventure/pull/950 last comment

File: front/src/WebRtc/MediaManager.ts [hunk 18]
```diff
@@ -405,33 +411,33 @@ export class MediaManager {
...
         element.classList.remove('active');
-        if(volume <= 0){
+        if (volume <= 0) {
             return;
         }
         element.classList.add('active');
         element.childNodes.forEach((value: ChildNode, index) => {
             const elementChildre = element.children.item(index);
-            if(!elementChildre){
+            if (!elementChildre) {
```
Comment (moufmouf) [inline front/src/WebRtc/MediaManager.ts:436]: Rename to `elementChildren`

File: front/src/Api/iframe/room.ts [hunk 19]
```diff
@@ -113,7 +113,7 @@ class WorkadventureRoomCommands extends IframeApiContribution<WorkadventureRoomC
                 'layerName': layerName,
                 'propertyName': propertyName,
                 'propertyValue': propertyValue,
-            } as SetPropertyEvent
```
Comment (jonnytest1) [inline front/src/Api/iframe/room.ts]: i mean you dont need to remove them all just saying its not completely neccessary :D

**Review by moufmouf — APPROVED** [review]
